product_weight/wizard/product_weight_update.py

```diff
@@ -0,0 +1,116 @@
+            return res
+        context = self.env.context
+        if context.get('active_model') == 'product.template':
+            product_tmpl_id = context.get('active_id', False)
+            domain_template = [('product_tmpl_id', '=', product_tmpl_id)]
+            domain_product = []
+        else:
+            product_id = context.get('active_id', False)
+            product = self.env['product.product'].browse(product_id)
+            product_tmpl_id = product.product_tmpl_id.id
+            domain_template = [('product_tmpl_id', '=', product_tmpl_id)]
+            domain_product = [('product_id', '=', product_id)]
+        boms = []
+        if domain_product:
+            boms = self.env['mrp.bom'].search(domain_product)
+        if not domain_product or not boms:
+            boms = self.env['mrp.bom'].search(domain_template)
+        if boms:
+            bom = boms[0]
+            res.update({'bom_id': bom.id})
+
+        if 'product_tmpl_id' in fields:
+            res.update({'product_tmpl_id': product_tmpl_id})
+        return res
+
+    @api.multi
+    def calculate_product_bom_weight(self, bom):
+        uom_obj = self.env['product.uom']
+        product_tmpl = bom.product_tmpl_id
+        tmpl_qty = uom_obj._compute_qty(
+            bom.product_uom.id,
+            bom.product_qty,
+            product_tmpl.uom_id.id)
+        bom_lines = bom.bom_line_ids.get_final_components()
+        weight_gross = 0.0
+        weight_net = 0.0
+        for line in bom_lines:
+            component_tmpl = line.product_id.product_tmpl_id
+            component_qty = uom_obj._compute_qty(
+                line.product_uom.id,
+                line.product_qty,
+                component_tmpl.uom_id.id)
+            weight_net += component_tmpl.weight_net * component_qty
+            weight_gross += component_tmpl.weight * component_qty
+            _logger.warning("%s : %0.2f | %0.2f" % (
+                bom.product_tmpl_id.name,
+                weight_net, weight_gross))
+        weight_net = weight_net / tmpl_qty
+        weight_gross = weight_gross / tmpl_qty
+        product_tmpl.write({'weight': weight_gross, 'weight_net': weight_net})
+
+    @api.multi
+    def update_single_weight(self):
+        self.ensure_one()
+        self.calculate_product_bom_weight(self.bom_id)
+        return {}
+
+    @api.multi
+    def update_multi_product_weight(self):
+        self.ensure_one()
+        product_obj = self.env['product.product']
+        context = self.env.context
+        if context.get('active_model') == 'product.template':
+            template_ids = context.get('active_ids', [])
+        else:
+            product_ids = context.get('active_ids', [])
+            products = product_obj.browse(product_ids)
+            template_ids = [p.product_tmpl_id.id for p in products]
+        for template_id in template_ids:
+            boms = self.env['mrp.bom'].search(
+                [('product_tmpl_id', '=', template_id)])
```

Comment:
You are looking for a single bom? You should limit your search with `limit=1` to assign `bom`